File: src/main/scala/resonantinduction/archaic/fluid/gutter/TileGutter.java
package resonantinduction.archaic.fluid.gutter;

import java.util.ArrayList;
import java.util.List;

import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.RenderBlocks;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.client.model.AdvancedModelLoader;
import net.minecraftforge.client.model.IModelCustom;
import net.minecraftforge.common.ForgeDirection;
import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidRegistry;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.FluidTank;
import net.minecraftforge.fluids.IFluidHandler;
import net.minecraftforge.fluids.IFluidTank;

import org.lwjgl.opengl.GL11;

import resonant.api.recipe.MachineRecipes;
import resonant.api.recipe.RecipeResource;
import resonant.lib.content.module.TileRender;
import resonant.lib.prefab.vector.Cuboid;
import resonant.lib.render.FluidRenderUtility;
import resonant.lib.render.RenderUtility;
import resonant.lib.utility.FluidUtility;
import resonant.lib.utility.WorldUtility;
import resonant.lib.utility.inventory.InventoryUtility;
import resonantinduction.archaic.fluid.grate.TileGrate;
import resonantinduction.core.Reference;
import resonantinduction.core.ResonantInduction.RecipeType;
import resonantinduction.core.grid.fluid.FluidPressureNode;
import resonantinduction.core.grid.fluid.IPressureNodeProvider;
import resonantinduction.core.grid.fluid.TilePressureNode;
import universalelectricity.api.vector.Vector3;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

/**
 * The gutter, used for fluid transfer.
 *
 * @author Calclavia
 */
public class TileGutter extends TilePressureNode
{
	public TileGutter()
	{
		super(Material.rock);
		textureName = "material_wood_surface";
		isOpaqueCube = false;
		normalRender = false;
		bounds = new Cuboid(0, 0, 0, 1, 0.99, 1);

		node = new FluidGravityNode(this)
		{
			@Override
			public void doRecache()
			{
				connections.clear();
				byte previousConnections = renderSides;
				renderSides = 0;

				for (ForgeDirection dir : ForgeDirection.VALID_DIRECTIONS)
				{
					TileEntity tile = position().translate(dir).getTileEntity(world());

					if (tile instanceof IFluidHandler)
					{
						if (tile instanceof IPressureNodeProvider)
						{
							FluidPressureNode check = (FluidPressureNode) ((IPressureNodeProvider) tile).getNode(FluidPressureNode.class, dir.getOpposite());

							if (check != null && canConnect(dir, check) && check.canConnect(dir.getOpposite(), this))
							{
								connections.put(check, dir);

								if (tile instanceof TileGutter)
								{
									renderSides = WorldUtility.setEnableSide(renderSides, dir, true);
								}

							}
						}
						else
						{
							connections.put(tile, dir);

							if (tile instanceof TileGrate)
							{
								renderSides = WorldUtility.setEnableSide(renderSides, dir, true);
							}
						}
					}
				}

				/** Only send packet updates if visuallyConnected changed. */
				if (previousConnections != renderSides)
				{
					sendRenderUpdate();
				}
			}
		};
	}

	@Override
	public void updateEntity()
	{
		super.updateEntity();

		if (!this.world().isRemote && this.ticks % 20 == 0)
		{
			/** Drain block above if it is a fluid. */
			Vector3 drainPos = position().translate(0, 1, 0);
			FluidStack drain = FluidUtility.drainBlock(worldObj, drainPos, false);

			if (drain != null)
			{
				ArrayList<IFluidTank> tanks = new ArrayList<IFluidTank>();

				synchronized (node.getGrid().getNodes())
				{
					for (Object check : node.getGrid().getNodes())
					{
						if (check instanceof FluidPressureNode)
						{
							if (((FluidPressureNode) check).parent instanceof TileGutter)
							{
								tanks.add(((FluidPressureNode) check).parent.getPressureTank());
							}
						}
					}
				}

				if (FluidUtility.fillAllTanks(tanks, drain, false) >= drain.amount)
				{
					FluidUtility.fillAllTanks(tanks, drain, true);
					FluidUtility.drainBlock(worldObj, drainPos, true);

					synchronized (node.getGrid().getNodes())
					{
						for (Object check : node.getGrid().getNodes())
						{
							if (check instanceof FluidPressureNode)
							{
								((FluidPressureNode) check).parent.onFluidChanged();
							}
						}
					}
				}
			}
		}
	}

	@Override
	public Iterable<Cuboid> getCollisionBoxes()
	{
		List<Cuboid> list = new ArrayList<Cuboid>();

		float thickness = 0.1F;

		if (!WorldUtility.isEnabledSide(renderSides, ForgeDirection.DOWN))
		{
			list.add(new Cuboid(0.0F, 0.0F, 0.0F, 1.0F, thickness, 1.0F));
		}

		if (!WorldUtility.isEnabledSide(renderSides, ForgeDirection.WEST))
		{
			list.add(new Cuboid(0.0F, 0.0F, 0.0F, thickness, 1.0F, 1.0F));
		}
		if (!WorldUtility.isEnabledSide(renderSides, ForgeDirection.NORTH))
		{
			list.add(new Cuboid(0.0F, 0.0F, 0.0F, 1.0F, 1.0F, thickness));
		}

		if (!WorldUtility.isEnabledSide(renderSides, ForgeDirection.EAST))
		{
			list.add(new Cuboid(1.0F - thickness, 0.0F, 0.0F, 1.0F, 1.0F, 1.0F));
		}

		if (!WorldUtility.isEnabledSide(renderSides, ForgeDirection.SOUTH))
		{
			list.add(new Cuboid(0.0F, 0.0F, 1.0F - thickness, 1.0F, 1.0F, 1.0F));
		}

		return list;
	}

	@Override
	public void collide(Entity entity)
	{
		if (getInternalTank().getFluidAmount() > 0)
		{
			for (int i = 2; i < 6; i++)
			{
				ForgeDirection dir = ForgeDirection.getOrientation(i);
				int pressure = node.getPressure(dir);
				Vector3 position = position().translate(dir);

				TileEntity checkTile = position.getTileEntity(world());

				if (checkTile instanceof TileGutter)
				{
					int deltaPressure = pressure - ((TileGutter) checkTile).node.getPressure(dir.getOpposite());

					entity.motionX += 0.01 * dir.offsetX * deltaPressure;
					entity.motionY += 0.01 * dir.offsetY * deltaPressure;
					entity.motionZ += 0.01 * dir.offsetZ * deltaPressure;
				}
			}

			if (getInternalTank().getFluid().getFluid().getTemperature() >= 373)
			{
				entity.setFire(5);
			}
			else
			{
				entity.extinguish();
			}
		}

		if (entity instanceof EntityItem)
		{
			entity.noClip = true;
		}
	}

	@Override
	public boolean activate(EntityPlayer player, int side, Vector3 vector3)
	{
		if (player.getCurrentEquippedItem() != null)
		{
			/** Manually wash dust into refined dust. */
			ItemStack itemStack = player.getCurrentEquippedItem();

			RecipeResource[] outputs = MachineRecipes.INSTANCE.getOutput(RecipeType.MIXER.name(), itemStack);

			if (outputs.length > 0)
			{
				if (!world().isRemote)
				{
					int drainAmount = 50 + world().rand.nextInt(50);
					FluidStack drain = drain(ForgeDirection.UP, drainAmount, false);

					if (drain != null && drain.amount > 0 && world().rand.nextFloat() > 0.9)
					{
						if (world().rand.nextFloat() > 0.1)
						{
							for (RecipeResource res : outputs)
							{
								InventoryUtility.dropItemStack(world(), new Vector3(player), res.getItemStack().copy(), 0);
							}
						}

						itemStack.stackSize--;

						if (itemStack.stackSize <= 0)
						{
							itemStack = null;
						}

						player.inventory.setInventorySlotContents(player.inventory.currentItem, itemStack);
					}

					drain(ForgeDirection.UP, drainAmount, true);

					world().playSoundEffect(x() + 0.5, y() + 0.5, z() + 0.5, "liquid.water", 0.5f, 1);
				}

				return true;
			}
		}

		if (!world().isRemote)
		{
			ArrayList<IFluidTank> tanks = new ArrayList<IFluidTank>();

			synchronized (node.getGrid().getNodes())
			{
				for (Object check : node.getGrid().getNodes())
				{
					if (check instanceof FluidPressureNode)
					{
						if (((FluidPressureNode) check).parent instanceof TileGutter)
						{
							tanks.add(((FluidPressureNode) check).parent.getPressureTank());
						}
					}
				}
			}

			if (FluidUtility.playerActivatedFluidItem(tanks, player, side))
			{
				synchronized (node.getGrid().getNodes())
				{
					for (Object check : node.getGrid().getNodes())
					{
						if (check instanceof FluidPressureNode)
						{
							((FluidPressureNode) check).parent.onFluidChanged();
						}
					}
				}

				return true;
			}

			return false;

		}

		return true;
	}

	@Override
	public void onFillRain()
	{
		if (!world().isRemote)
		{
			fill(ForgeDirection.UP, new FluidStack(FluidRegistry.WATER, 10), true);
		}
	}

	@Override
	public int fill(ForgeDirection from, FluidStack resource, boolean doFill)
	{
		if (!resource.getFluid().isGaseous())
		{
			return super.fill(from, resource, doFill);
		}

		return 0;
	}

	@Override
	public boolean canFill(ForgeDirection from, Fluid fluid)
	{
		return from != ForgeDirection.UP && !fluid.isGaseous();
	}

	@Override
	public boolean canDrain(ForgeDirection from, Fluid fluid)
	{
		return from != ForgeDirection.UP && !fluid.isGaseous();
	}

	@SideOnly(Side.CLIENT)
	@Override
	protected TileRender newRenderer()
	{
		return new TileRender()
		{
			public final IModelCustom MODEL = AdvancedModelLoader.loadModel(Reference.MODEL_DIRECTORY + "gutter.tcn");
			public final ResourceLocation TEXTURE = new ResourceLocation(Reference.DOMAIN, Reference.MODEL_PATH + "gutter.png");

			@Override
			public boolean renderStatic(RenderBlocks renderer, Vector3 position)
			{
				return true;
			}

			@Override
			public boolean renderDynamic(Vector3 position, boolean isItem, float frame)
			{
				GL11.glPushMatrix();
				GL11.glTranslated(position.x + 0.5, position.y + 0.5, position.z + 0.5);

				FluidStack liquid = getInternalTank().getFluid();
				int capacity = getInternalTank().getCapacity();

				render(0, renderSides);

				if (world() != null)
				{
					FluidTank tank = getInternalTank();
					double percentageFilled = (double) tank.getFluidAmount() / (double) tank.getCapacity();

					if (percentageFilled > 0.1)
					{
						GL11.glPushMatrix();
						GL11.glScaled(0.990, 0.99, 0.990);

						double ySouthEast = FluidUtility.getAveragePercentageFilledForSides(TileGutter.class, percentageFilled, world(), position(), ForgeDirection.SOUTH, ForgeDirection.EAST);
						double yNorthEast = FluidUtility.getAveragePercentageFilledForSides(TileGutter.class, percentageFilled, world(), position(), ForgeDirection.NORTH, ForgeDirection.EAST);
						double ySouthWest = FluidUtility.getAveragePercentageFilledForSides(TileGutter.class, percentageFilled, world(), position(), ForgeDirection.SOUTH, ForgeDirection.WEST);
						double yNorthWest = FluidUtility.getAveragePercentageFilledForSides(TileGutter.class, percentageFilled, world(), position(), ForgeDirection.NORTH, ForgeDirection.WEST);

						FluidRenderUtility.renderFluidTesselation(tank, ySouthEast, yNorthEast, ySouthWest, yNorthWest);
						GL11.glPopMatrix();
					}
				}

				GL11.glPopMatrix();
				return true;
			}

			@Override
			public boolean renderItem(ItemStack itemStack)
			{
				GL11.glTranslated(0.5, 0.5, 0.5);
				render(itemStack.getItemDamage(), Byte.parseByte("001100", 2));
				return true;
			}

			public void render(int meta, byte sides)
			{
				RenderUtility.bind(TEXTURE);

				double thickness = 0.055;
				double height = 0.5;

				for (ForgeDirection dir : ForgeDirection.VALID_DIRECTIONS)
				{
					if (dir != ForgeDirection.UP && dir != ForgeDirection.DOWN)
					{
						GL11.glPushMatrix();
						RenderUtility.rotateBlockBasedOnDirection(dir);

						if (WorldUtility.isEnabledSide(sides, ForgeDirection.DOWN))
						{
							GL11.glTranslatef(0, -0.075f, 0);
							GL11.glScalef(1, 1.15f, 1);
						}

						if (!WorldUtility.isEnabledSide(sides, dir))
						{
							/** Render sides */
							MODEL.renderOnly("left");
						}

						if (!WorldUtility.isEnabledSide(sides, dir) || !WorldUtility.isEnabledSide(sides, dir.getRotation(ForgeDirection.UP)))
						{
							/** Render strips */
							MODEL.renderOnly("backCornerL");
						}
						GL11.glPopMatrix();
					}
				}

				if (!WorldUtility.isEnabledSide(sides, ForgeDirection.DOWN))
				{
					MODEL.renderOnly("base");
				}

			}
		};
	}

}

File: src/main/scala/resonantinduction/archaic/fluid/gutter/TileGutter.scala
package resonantinduction.archaic.fluid.gutter

import java.util.{ArrayList, List}

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.model.AdvancedModelLoader
import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.fluids._
import org.lwjgl.opengl.GL11
import resonant.api.recipe.MachineRecipes
import resonant.lib.factory.resources.RecipeType
import resonant.lib.prefab.fluid.NodeFluid
import resonant.lib.render.{FluidRenderUtility, RenderUtility}
import resonant.lib.transform.region.Cuboid
import resonant.lib.transform.vector.Vector3
import resonant.lib.utility.FluidUtility
import resonant.lib.utility.inventory.InventoryUtility
import resonant.lib.wrapper.BitmaskWrapper._
import resonantinduction.archaic.fluid.tank.TileTank
import resonantinduction.core.Reference
import resonantinduction.core.prefab.node.{NodeFluidPressure, TileFluidProvider}

import scala.collection.convert.wrapAll._

object TileGutter
{
  @SideOnly(Side.CLIENT)
  private val model = AdvancedModelLoader.loadModel(new ResourceLocation(Reference.domain, Reference.modelPath + "gutter.tcn"))
  @SideOnly(Side.CLIENT)
  private val texture = new ResourceLocation(Reference.domain, Reference.modelPath + "gutter.png")
}

/**
 * The gutter, used for fluid transfer.
 *
 * @author Calclavia
 */
class TileGutter extends TileFluidProvider(Material.rock)
{
  fluidNode = new NodeGutter(this)
  fluidNode.asInstanceOf[NodeGutter].doPressureUpdate = false

  fluidNode.onConnectionChanged = () =>
  {
    sendPacket(0)
  }

  fluidNode.onFluidChanged = () =>
  {
    if (!world.isRemote)
    {
      //TODO: Check if this is very costly
      //      UpdateTicker.world.enqueue(checkFluidAbove)
      sendPacket(1)
    }
  }

  textureName = "material_wood_surface"
  isOpaqueCube = false
  normalRender = false
  bounds = new Cuboid(0, 0, 0, 1, 0.99, 1)

  override def getCollisionBoxes: java.lang.Iterable[Cuboid] =
  {
    val list: List[Cuboid] = new ArrayList[Cuboid]
    val thickness = 0.1f

    if (!clientRenderMask.mask(ForgeDirection.DOWN))
    {
      list.add(new Cuboid(0.0F, 0.0F, 0.0F, 1.0F, thickness, 1.0F))
    }
    if (!clientRenderMask.mask(ForgeDirection.WEST))
    {
      list.add(new Cuboid(0.0F, 0.0F, 0.0F, thickness, 1.0F, 1.0F))
    }
    if (!clientRenderMask.mask(ForgeDirection.NORTH))
    {
      list.add(new Cuboid(0.0F, 0.0F, 0.0F, 1.0F, 1.0F, thickness))
    }
    if (!clientRenderMask.mask(ForgeDirection.EAST))
    {
      list.add(new Cuboid(1.0F - thickness, 0.0F, 0.0F, 1.0F, 1.0F, 1.0F))
    }
    if (!clientRenderMask.mask(ForgeDirection.SOUTH))
    {
      list.add(new Cuboid(0.0F, 0.0F, 1.0F - thickness, 1.0F, 1.0F, 1.0F))
    }
    return list
  }

  override def collide(entity: Entity)
  {
    if (fluidNode.getFluidAmount > 0)
    {
      for (i <- 2 to 6)
      {
        val dir: ForgeDirection = ForgeDirection.getOrientation(i)
        val pressure: Int = fluidNode.asInstanceOf[NodeFluidPressure].pressure(dir)
        val pos: Vector3 = toVector3.add(dir)
        val checkTile: TileEntity = pos.getTileEntity(world)

        if (checkTile.isInstanceOf[TileGutter])
        {
          val deltaPressure: Int = pressure - checkTile.asInstanceOf[TileGutter].fluidNode.asInstanceOf[NodeFluidPressure].pressure(dir.getOpposite)
          entity.motionX += 0.01 * dir.offsetX * deltaPressure
          entity.motionY += 0.01 * dir.offsetY * deltaPressure
          entity.motionZ += 0.01 * dir.offsetZ * deltaPressure
        }
      }
      if (fluidNode.getFluid.getFluid.getTemperature >= 373)
      {
        entity.setFire(5)
      }
      else
      {
        entity.extinguish()
      }
    }
    if (entity.isInstanceOf[EntityItem])
    {
      entity.noClip = true
    }
  }

  /**
   * Called when the block is first added to the world
   */
  override def onWorldJoin()
  {
    super.onWorldJoin()
    checkFluidAbove()
  }

  override def onNeighborChanged(block: Block)
  {
    super.onNeighborChanged(block)
    checkFluidAbove()
  }

  def checkFluidAbove()
  {
    if (!world.isRemote)
    {
      val posAbove = toVectorWorld + new Vector3(0, 1, 0)
      val blockAbove = posAbove.getBlock
      val tanks = findAllTanks

      def drainAbove(doDrain: Boolean) =
      {
        val stack = FluidUtility.drainBlock(posAbove.world, posAbove, doDrain)

        if (stack != null)
          FluidUtility.fillAllTanks(tanks, stack, doDrain)
        else
          0
      }

      if (drainAbove(false) == FluidContainerRegistry.BUCKET_VOLUME)
        drainAbove(true)
    }
  }

  override def use(player: EntityPlayer, side: Int, vector3: Vector3): Boolean =
  {
    if (player.getCurrentEquippedItem != null)
    {
      val itemStack = player.getCurrentEquippedItem
      val outputs = MachineRecipes.instance.getOutput(RecipeType.MIXER.name, itemStack)

      if (outputs.length > 0)
      {
        if (!world.isRemote)
        {
          val drainAmount = 25
          val drain = fluidNode.drain(ForgeDirection.UP, drainAmount, false)

          if (drain != null && drain.amount > 0 && world.rand.nextFloat < 0.05)
          {
            outputs.map(_.getItemStack.copy).foreach(s => InventoryUtility.dropItemStack(world, new Vector3(player), s, 0))
            itemStack.stackSize -= 1
            if (itemStack.stackSize <= 0)
              player.inventory.setInventorySlotContents(player.inventory.currentItem, null)
          }
          player.addExhaustion(0.1f)
          fluidNode.drain(ForgeDirection.UP, drainAmount, true)
        }

        world.playSoundEffect(x + 0.5, y + 0.5, z + 0.5, "liquid.water", 0.5f, 1)
        return true
      }

      return FluidUtility.playerActivatedFluidItem(findAllTanks, player, side)
    }
    return true
  }

  override def onFillRain()
  {
    if (!world.isRemote)
    {
      fill(ForgeDirection.UP, new FluidStack(FluidRegistry.WATER, 10), true)
    }
  }

  @SideOnly(Side.CLIENT)
  override def renderInventory(itemStack: ItemStack)
  {
    render(0, 0x0)
  }

  override def renderDynamic(pos: Vector3, frame: Float, pass: Int)
  {
    GL11.glPushMatrix()
    GL11.glTranslated(pos.x + 0.5, pos.y + 0.5, pos.z + 0.5)

    render(0, clientRenderMask)

    if (world != null)
    {
      val tank: IFluidTank = fluidNode
      val percentageFilled = tank.getFluidAmount / tank.getCapacity.toDouble

      if (percentageFilled > 0.1)
      {
        GL11.glPushMatrix()
        GL11.glScaled(0.99, 0.99, 0.99)
        val ySouthEast = FluidUtility.getAveragePercentageFilledForSides(classOf[TileGutter], percentageFilled, world, toVectorWorld, ForgeDirection.SOUTH, ForgeDirection.EAST)
        val yNorthEast = FluidUtility.getAveragePercentageFilledForSides(classOf[TileGutter], percentageFilled, world, toVectorWorld, ForgeDirection.NORTH, ForgeDirection.EAST)
        val ySouthWest = FluidUtility.getAveragePercentageFilledForSides(classOf[TileGutter], percentageFilled, world, toVectorWorld, ForgeDirection.SOUTH, ForgeDirection.WEST)
        val yNorthWest = FluidUtility.getAveragePercentageFilledForSides(classOf[TileGutter], percentageFilled, world, toVectorWorld, ForgeDirection.NORTH, ForgeDirection.WEST)
        FluidRenderUtility.renderFluidTesselation(tank, ySouthEast, yNorthEast, ySouthWest, yNorthWest)
        GL11.glPopMatrix()
      }
    }

    GL11.glPopMatrix()
  }

  def render(meta: Int, sides: Int)
  {
    RenderUtility.bind(TileGutter.texture)

    for (dir <- ForgeDirection.VALID_DIRECTIONS)
    {
      if (dir != ForgeDirection.UP && dir != ForgeDirection.DOWN)
      {
        GL11.glPushMatrix()
        RenderUtility.rotateBlockBasedOnDirection(dir)

        if (sides.mask(ForgeDirection.DOWN))
        {
          GL11.glTranslatef(0, -0.075f, 0)
          GL11.glScalef(1, 1.15f, 1)
        }
        if (!sides.mask(dir))
        {
          TileGutter.model.renderOnly("left")
        }
        if (!sides.mask(dir) || !sides.mask(dir.getRotation(ForgeDirection.UP)))
        {
          TileGutter.model.renderOnly("backCornerL")
        }
        GL11.glPopMatrix()
      }
    }

    if (!sides.mask(ForgeDirection.DOWN))
    {
      TileGutter.model.renderOnly("base")
    }
  }

  //Recurse through all gutter blocks
  private def findGutters(traverse: Set[NodeGutter] = Set(fluidNode.asInstanceOf[NodeGutter])): Set[NodeGutter] =
  {
    val current = traverse.last
    var foundGutters = traverse

    current.connections
      .filter(g => g.isInstanceOf[NodeGutter] && !traverse.contains(g))
      .map(_.asInstanceOf[NodeGutter])
      .foreach(g => foundGutters ++= findGutters(traverse + g))

    return foundGutters
  }

  private def findAllTanks = findGutters().map(_.getPrimaryTank).toList

  class NodeGutter(parent: TileFluidProvider) extends NodeFluidGravity(parent)
  {

    override def connect[B <: IFluidHandler](obj: B, dir: ForgeDirection) =
    {
      super.connect(obj, dir)

      if (obj.isInstanceOf[NodeGutter])
        clientRenderMask = clientRenderMask.openMask(dir)
    }

    override def clearConnections()
    {
      super.clearConnections()
      clientRenderMask = 0x00
    }

    override def canConnect[B <: IFluidHandler](other: B, from: ForgeDirection): Boolean =
    {
      if (other.isInstanceOf[NodeFluid] && other.asInstanceOf[NodeFluid].parent.isInstanceOf[TileTank])
        return false

      return super.canConnect(other, from)
    }

    override def canFill(from: ForgeDirection, fluid: Fluid): Boolean =
    {
      return from != ForgeDirection.UP && !fluid.isGaseous
    }

    override def canDrain(from: ForgeDirection, fluid: Fluid): Boolean =
    {
      return from != ForgeDirection.UP && !fluid.isGaseous
    }

    override def fill(from: ForgeDirection, resource: FluidStack, doFill: Boolean): Int =
    {
      if (!resource.getFluid.isGaseous)
      {
        return super.fill(from, resource, doFill)
      }
      return 0
    }
  }

}